File: plugins/org.projectforge.plugins.financialfairplay/src/main/java/org/projectforge/plugins/ffp/model/FFPAccountingDO.java
package org.projectforge.plugins.ffp.model;

import java.math.BigDecimal;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.Table;

import org.projectforge.business.fibu.EmployeeDO;
import org.projectforge.common.anots.PropertyInfo;
import org.projectforge.framework.persistence.entities.DefaultBaseDO;

import de.micromata.genome.db.jpa.history.api.WithHistory;

@Entity
@Table(name = "T_PLUGIN_FINANCIALFAIRPLAY_ACCOUNTING")
@WithHistory
public class FFPAccountingDO extends DefaultBaseDO
{
  private static final long serialVersionUID = -1361266966025898919L;

  private FFPEventDO event;

  private EmployeeDO attendee;

  @PropertyInfo(i18nKey = "plugins.ffp.value")
  private BigDecimal value;

  @PropertyInfo(i18nKey = "plugins.ffp.weighting")
  private BigDecimal weighting;

  @PropertyInfo(i18nKey = "plugins.ffp.comment")
  private String comment;

  @ManyToOne(fetch = FetchType.EAGER)
  @JoinColumn(name = "EVENT_ID")
  public FFPEventDO getEvent()
  {
    return event;
  }

  public void setEvent(FFPEventDO event)
  {
    this.event = event;
  }

  @ManyToOne(fetch = FetchType.EAGER)
  @JoinColumn(name = "ATTENDEE_ID")
  public EmployeeDO getAttendee()
  {
    return attendee;
  }

  public void setAttendee(EmployeeDO attendee)
  {
    this.attendee = attendee;
  }

  @Column(nullable = false)
  public BigDecimal getValue()
  {
    return value;
  }

  public void setValue(BigDecimal value)
  {
    this.value = value;
  }

  @Column(nullable = false)
  public BigDecimal getWeighting()
  {
    return weighting;
  }

  public void setWeighting(BigDecimal weighting)
  {
    this.weighting = weighting;
  }

  @Column
  public String getComment()
  {
    return comment;
  }

  public void setComment(String comment)
  {
    this.comment = comment;
  }

  @Override
  public boolean equals(Object o)
  {
    if (o instanceof FFPAccountingDO == false) {
      return false;
    }
    FFPAccountingDO other = (FFPAccountingDO) o;
    if (this.getId() != null && other.getId() != null) {
      return this.getId().equals(other.getId());
    }
    if (this.getEvent() != null && this.getEvent().getId() != null && other.getEvent() != null && other.getEvent().getId() != null
        && this.getAttendee() != null && other.getAttendee() != null) {
      return this.getEvent().getId().equals(other.getEvent().getId())
          && this.getAttendee().getId().equals(other.getAttendee().getId());
    }
    //Case new event
    if (this.getAttendee() != null && other.getAttendee() != null) {
      return this.getAttendee().getId().equals(other.getAttendee().getId());
    }
    return false;
  }

  @Override
  public int hashCode()
  {
    int result = (getId() != null) ? getId().hashCode() : 0;
    result = 31 * result + (event != null && event.getId() != null ? event.getId().hashCode() : 0);
    result = 31 * result + (attendee != null ? attendee.getId().hashCode() : 0);
    return result;
  }
}

File: plugins/org.projectforge.plugins.financialfairplay/src/main/kotlin/org/projectforge/plugins/ffp/model/FFPAccountingDO.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2019 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.plugins.ffp.model

import de.micromata.genome.db.jpa.history.api.WithHistory
import org.projectforge.common.anots.PropertyInfo
import org.projectforge.framework.persistence.entities.DefaultBaseDO
import org.projectforge.framework.persistence.user.entities.PFUserDO
import java.math.BigDecimal
import javax.persistence.*

@Entity
@Table(name = "T_PLUGIN_FINANCIALFAIRPLAY_ACCOUNTING")
@WithHistory
open class FFPAccountingDO : DefaultBaseDO() {

    @get:ManyToOne(fetch = FetchType.EAGER)
    @get:JoinColumn(name = "EVENT_ID")
    open var event: FFPEventDO? = null

    @get:ManyToOne(fetch = FetchType.EAGER)
    @get:JoinColumn(name = "ATTENDEE_USER_ID")
    open var attendee: PFUserDO? = null

    @PropertyInfo(i18nKey = "plugins.ffp.value")
    @get:Column(nullable = false)
    open var value: BigDecimal? = null

    @PropertyInfo(i18nKey = "plugins.ffp.weighting")
    @get:Column(nullable = false)
    open var weighting: BigDecimal? = null

    @PropertyInfo(i18nKey = "plugins.ffp.comment")
    @get:Column
    open var comment: String? = null

    override fun equals(other: Any?): Boolean {
        if (other !is FFPAccountingDO) {
            return false
        }
        val o = other as FFPAccountingDO?
        if (this.id != null && o!!.id != null) {
            return this.id == o.id
        }
        if (this.event != null && this.event!!.id != null && o!!.event != null && o.event!!.id != null
                && this.attendee != null && o.attendee != null) {
            return this.event!!.id == o.event!!.id && this.attendee!!.id == o.attendee!!.id
        }
        //Case new event
        return if (this.attendee != null && o!!.attendee != null) {
            this.attendee!!.id == o.attendee!!.id
        } else false
    }

    override fun hashCode(): Int {
        var result = if (id != null) id!!.hashCode() else 0
        result = 31 * result + if (event != null && event!!.id != null) event!!.id!!.hashCode() else 0
        result = 31 * result + if (attendee != null) attendee!!.id!!.hashCode() else 0
        return result
    }
}
